cluster/swarm/cluster.go

```diff
@@ -359,9 +359,13 @@ func (c *Cluster) removeEngine(addr string) bool {
 func (c *Cluster) monitorDiscovery(ch <-chan discovery.Entries, errCh <-chan error) {
 	// Watch changes on the discovery channel.
 	currentEntries := discovery.Entries{}
-	for {
+	for ch != nil || errCh != nil {
 		select {
-		case entries := <-ch:
+		case entries, ok := <-ch:
+			if !ok {
+				ch = nil
```

Comment:
@dongluochen 
> The reason here is I don't think monitorDiscovery should change the channels. It was not designed to do that.

Very sorry for my stubbornness. I think `monitorDiscovery` doesn't **change** the channels, and it just set local channel variable to  `nil` when it knows the channel is closed.